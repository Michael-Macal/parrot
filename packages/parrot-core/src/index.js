export validateScenarios from './validateScenarios';
export normalizeScenarios from './normalizeScenarios';
export matchMock from './matchMock';
export resolveResponse from './resolveResponse';
export logger from './utils/logger';
