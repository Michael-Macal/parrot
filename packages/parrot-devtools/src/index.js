import LocaleSelector from './components/LocaleSelector';
import ProductSelector from './components/ProductSelector';
import ScenarioSelector from './components/ScenarioSelector';

export {
  LocaleSelector,
  ProductSelector,
  ScenarioSelector
};
export default ParrotCage;
