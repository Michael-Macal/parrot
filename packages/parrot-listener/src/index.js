export default from './parrotListener';
export setupCmdArgs from './setupCmdArgs';
