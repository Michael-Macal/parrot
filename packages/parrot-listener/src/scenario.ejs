export default {
  <%- name %>: [<% routes.forEach(function(route) { %>{
      request: <%- JSON.stringify(route.request) %>,
      response: require('<%- route.response %>')
    },
  <% }); %>]
};
