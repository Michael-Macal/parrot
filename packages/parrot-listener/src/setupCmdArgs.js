const setupCmdArgs = (yargs) =>
  yargs
  .option('l', {
     alias: 'listen',
     describe: 'Listen to incoming requests and generate scenarios',
     type: 'boolean',
   })
   .option('n', {
     alias: 'name',
     describe: 'Name for your generated scenario',
     type: 'string',
   });
   // TODO implement overwriting
   /*.option('update', {
     alias: 'u',
     describe: 'Overwrite your scenario if it exists',
     type: 'boolean',
   });*/

export default setupCmdArgs;
