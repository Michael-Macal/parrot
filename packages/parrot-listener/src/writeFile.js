import mkdirp from 'mkdirp';
import path from 'path';
import promisify from 'promisify-node';

// Wraps fs.writeFile to add handling for non-existing directories
const writeFile = (pathToFile, contents, writeAttempted = false) => {
  const fs = promisify('fs');
  return fs
    .writeFile(pathToFile, contents)
    .catch(err => {
      if (err.code === 'ENOENT' && !writeAttempted) {
        mkdirp(path.parse(pathToFile).dir, (err) => {
          if (err) {
            throw err;
          } else {
            return writeFile(pathToFile, contents, true);
          }
        });
      } else {
        throw err;
      }
    });
};

export default writeFile;
