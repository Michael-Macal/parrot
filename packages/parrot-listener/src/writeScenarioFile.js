import ejs from 'ejs';
import path from 'path';
import promisify from 'promisify-node';
import writeFile from './writeFile';

const writeScenarioFile = (name, routes, outputDir) => {
  const fs = promisify('fs');
  const templateString = fs.readFileSync(path.join(__dirname, './scenario.ejs'), 'utf-8');
  const contents = ejs.render(templateString, {
    name,
    routes,
  });
  return writeFile(path.join(__dirname, outputDir, name, 'scenario.js'), contents);
};

export default writeScenarioFile;
