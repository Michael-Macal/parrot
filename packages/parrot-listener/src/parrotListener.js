import bodyParser from 'body-parser';
import mkdirp from 'mkdirp';
import path from 'path';
import promisify from 'promisify-node';
import shortid from 'shortid';
import registerMiddleware from 'parrot-registry';
import writeFile from './writeFile';
import writeScenarioFile from './writeScenarioFile';

const fs = promisify('fs');

const parrotListener = ({
  output = '',
  matcher = () => true,
  name,
  parser = req => req,
  listening = false,
  logger = console.log,
}) => {
  let routes = [];
  let scenarioName = name;
  let isListening = listening;

  const startListening = (name) => {
    scenarioName = name;
    isListening = true;
  }

  const stopListening = () => {
    isListening = false;
    return writeScenarioFile(scenarioName, routes, output)
      .then(() => { routes = []; })
      .catch(logger);
  };

  const middleware = (req, res, next) => {
    if (matcher(req)) {
      res.locals.parrot = {
        id: shortid.generate(),
      };

      const chunks = [];
      const oldWrite = res.write;

      res.write = function (chunk) {
        chunks.push(chunk);
        oldWrite.apply(res, arguments);
      };

      res.on('finish', () => {
        try {
          if (res.statusCode >= 200 && res.statusCode < 300) {
            const filename = `${res.locals.parrot.id}.json`;
            const outputPath = path.join(__dirname, output, scenarioName, filename);
            const body = Buffer.concat(chunks).toString('utf8');

            return writeFile(outputPath, body)
              .then(() => {
                logger(`Created file for response.
                  request: ${req.path},
                  location: ${outputPath}
                `);
              })
              .then(() => {
                routes.unshift({
                  request: parser(req),
                  response: `./${scenarioName}/${filename}`,
                });
              })
              .catch((e) => {
                logger(`ERROR: Unable to write parrot-listener scenario file: ${e}\n`)
              });
          }
        } catch (e) {
          logger(`ERROR: Error in parrot-listener middleware: ${e.stack}`);
        }
      });

      const killListener = () => {
        if (isListening) {
          stopListening();
        }
      };
      process.on('exit', killListener);
      process.on('SIGINT', killListener);
    }
    next();
  };

  return (app) => {
    app.use(bodyParser.json());
    app.get('/parrot/listen', (req, res) => {
      const config = isListening ? { isListening, scenarioName } : { isListening };
      res.json(config);
    });

    app.put('/parrot/listen', (req, res) => {
      if (req.body.action === 'START') {
        if (isListening === true) {
          return res.status(500).json({
            reason: 'Already listening!'
          });
        } else if (!req.body.scenarioName) {
          return res.status(500).json({
            reason: 'Missing request field: "scenarioName"'
          });
        }
        startListening(req.body.scenarioName);
        return res.status(200).send();
      } else if (req.body.action === 'STOP') {
        return stopListening().then(() =>
          res.status(200).send()
        );
      } else {
        return res.status(500).json({
          reason: 'Unknown or missing action. Valid actions are: START, STOP'
        });
      }
    });
    app.use(middleware.bind(this));
    registerMiddleware(app, { name: 'parrot-listener' });
  };
}

export default parrotListener;
