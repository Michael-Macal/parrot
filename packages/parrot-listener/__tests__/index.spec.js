import * as API from '../src';

describe('API', () => {
  it('exports correct API', () => {
    expect(API).toMatchSnapshot();
  });
});
