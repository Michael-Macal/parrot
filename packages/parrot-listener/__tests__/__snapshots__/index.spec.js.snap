// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API exports correct API 1`] = `
Object {
  "default": [Function],
  "setupCmdArgs": [Function],
}
`;
